src/entities/pools/weighted/weightedJoin.ts

```diff
@@ -0,0 +1,226 @@
+import { createPublicClient, encodeFunctionData, http } from 'viem';
+import {
+    BaseJoin,
+    JoinCallInput,
+    JoinInput,
+    JoinQueryResult,
+    PoolState,
+} from '..';
+import { Address } from '../../../types';
+import {
+    BALANCER_HELPERS,
+    BALANCER_VAULT,
+    CHAINS,
+    ZERO_ADDRESS,
+} from '../../../utils';
+import { WeightedEncoder } from '../../encoders/weighted';
+import { TokenAmount } from '../../tokenAmount';
+import { balancerHelpersAbi } from '../../../abi/balancerHelpers';
+import { Token } from '../../token';
+import { vaultAbi } from '../../../abi';
+
+export class WeightedJoin implements BaseJoin {
+    // TODO - Probably not needed
+    getInstance(): WeightedJoin {
+        return new WeightedJoin();
+    }
+
+    public async query(
+        input: JoinInput,
+        poolState: PoolState,
+    ): Promise<JoinQueryResult> {
+        // TODO - This would need extended to work with relayer
+
+        this.checkInputs(input, poolState);
+
+        // infer join kind by input tokens
+        const joinKind = this.getJoinKind(input, poolState.address);
+
+        // Initialize join parameters
+        let maxAmountsIn = Array(poolState.tokens.length).fill(0n);
+        let userData: Address;
+        const poolAssets = poolState.tokens.map((t) => t.address);
+
+        switch (joinKind) {
+            case 'Init': {
+                maxAmountsIn = this.getAmountsIn(input, poolAssets);
+                userData = WeightedEncoder.joinInit(maxAmountsIn);
+                break;
+            }
+            case 'GivenIn': {
+                maxAmountsIn = this.getAmountsIn(input, poolAssets);
+                const bptOut = 0n;
+                userData = WeightedEncoder.joinGivenIn(maxAmountsIn, bptOut);
+                break;
+            }
+            case 'GivenOut': {
+                const bptOut = input.tokenAmounts[0].amount;
+                userData = WeightedEncoder.joinGivenOut(bptOut);
+                break;
+            }
+            default:
+                throw new Error('Invalid join kind');
+        }
+
+        const queryArgs = this.getJoinParameters({
+            poolId: poolState.id,
+            assets: poolAssets,
+            sender: ZERO_ADDRESS,
+            recipient: ZERO_ADDRESS,
+            maxAmountsIn,
+            userData,
+        });
+
+        const client = createPublicClient({
+            transport: http(input.rpcUrl),
```

Comment:
We don’t perform any network requests on client init.

Reply:
By checking this is not the case with `viem` you mean that we should omit this input?
I'll make it optional in case someone wants the easy path of using default values, but we also allow for custom rpcUrls just in case. 
Let me know if you think it's better to avoid that complexity at all.